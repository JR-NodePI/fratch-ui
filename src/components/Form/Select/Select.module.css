.select {
  font-family: var(--ft-select-default-font-family);
  font-size: var(--ft-select-default-font-size);
  box-shadow: var(--ft-box-shadow-1);
  border-radius: var(--ft-form-item-border-radius);
}
.select * {
  font-family: var(--ft-select-default-font-family);
  font-size: var(--ft-select-default-font-size);
}
.select.disabled * {
  pointer-events: none;
}

/*** cleaner ***/
.cleaner {
  border-radius: 0 var(--ft-form-item-border-radius)
    var(--ft-form-item-border-radius) 0;
  border: var(--ft-form-item-border);
  margin: 0;
  padding: var(--ft-form-item-padding-y) var(--ft-form-item-padding-x);
  border-left: none;
  transition: background-color 0.2s ease-out;
  background-color: var(--ft-button-default-bg-color);
}
.cleaner:hover {
  background-color: var(--ft-button-hover-bg-color);
}
.cleaner:active {
  background-color: var(--ft-button-active-bg-color);
}
.cleaner_icon [fill] {
  fill: var(--ft-button-default-color);
}
.cleaner .cleaner_icon [stroke] {
  stroke: var(--ft-button-default-color);
}
.cleaner .cleaner:hover .cleaner_icon [fill] {
  fill: var(--ft-button-hover-color);
}
.cleaner:hover .cleaner_icon [stroke] {
  stroke: var(--ft-button-hover-color);
}
.cleaner .cleaner_icon [fill='none'] {
  fill: transparent !important;
}
.cleaner .cleaner_icon [stroke='none'] {
  stroke: transparent !important;
}

.controls .trigger_icon,
.cleaner .cleaner_icon {
  font-size: var(--ft-button-default-font-size);
  margin: -0.1em;
  font-size: 1.1em;
  min-width: 1.1em;
}

.controls {
  display: flex;
  position: relative;
  flex-wrap: nowrap;
}
.controls > *:first-child {
  width: 100%;
  position: relative;
}
.controls .trigger_icon {
  margin-top: 0;
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  pointer-events: none;
  right: var(--ft-form-item-padding-x);
}
.controls .trigger {
  box-shadow: none;
}
.cleanable .trigger {
  border-radius: var(--ft-form-item-border-radius) 0 0
    var(--ft-form-item-border-radius);
}

.controls .trigger_icon [fill] {
  fill: var(--ft-form-item-border-color);
}
.controls .trigger_icon [stroke] {
  stroke: var(--ft-form-item-border-color);
}
.controls .trigger:hover ~ .trigger_icon [fill] {
  fill: var(--ft-color-primary);
}
.controls .trigger:hover ~ .trigger_icon [stroke] {
  stroke: var(--ft-color-primary);
}
.controls .trigger_icon [fill='none'] {
  fill: transparent !important;
}
.controls .trigger_icon [stroke='none'] {
  stroke: transparent !important;
}
