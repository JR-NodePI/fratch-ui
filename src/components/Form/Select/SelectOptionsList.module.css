.select_list {
  position: fixed;
  overflow: auto;
  z-index: var(--ft-max-z-index);
  background-color: var(--ft-form-item-bg-color);
  border-radius: var(--ft-form-item-border-radius);
  box-shadow: var(--ft-box-shadow-2);
  border: var(--ft-form-item-border);
  box-sizing: border-box;
}
.select_list ul {
  margin: 0;
  padding: 0;
}
.select_list ul li {
  list-style: none;
  margin: 0;
  padding: 0;
  position: relative;
  display: flex;
  flex-wrap: nowrap;
  min-width: 1em;
  box-sizing: border-box;
  box-shadow: inset 0 -1px 0 0 var(--ft-form-item-border-color-light);
  padding-bottom: 1px;
  overflow: hidden;
}
.select_list ul li:last-child {
  box-shadow: none;
}

.select_list ul li > span,
.select_list ul li > a {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  display: block;
  width: 100%;
  text-align: left;
  background-color: transparent;
  border: none;
  padding: var(--ft-form-item-padding-y) var(--ft-form-item-padding-x);
  margin: 0;
  text-decoration: none;
  outline: none;
  cursor: default;
}
.select_list ul li > span {
  color: var(--ft-form-item-border-color);
  background-color: var(--ft-primary-default-color);
}
.select_list ul li > a {
  color: var(--ft-button-default-color);
  background-color: var(--ft-form-item-bg-color);
}

.select_list ul li > a b {
  font-weight: 500;
  color: var(--ft-select-list-item-selected-text-color);
}
.select_list ul li > a * {
  pointer-events: none;
}

.select_list ul li > a:active,
.select_list ul li.selected > a {
  outline: none;
  background-color: var(--ft-select-list-item-bg-color-active);
}
.select_list ul li > a:focus,
.select_list ul li.focused > a {
  outline: none;
  background-color: var(--ft-select-list-item-bg-color-hover);
}

.select_list ul li.selected.focused > a {
  background-color: var(--ft-select-list-item-bg-color-hover-active);
}

.select_list.hidden {
  opacity: 0;
  transition: opacity var(--ft-transition-duration) ease-out,
    top 0s linear var(--ft-transition-duration),
    left 0s linear var(--ft-transition-duration),
    width 0s linear var(--ft-transition-duration),
    height 0s linear var(--ft-transition-duration),
    max-height 0s linear var(--ft-transition-duration);
}
.select_list.visible {
  opacity: 1;
  transition: opacity var(--ft-transition-duration) ease-out;
}
