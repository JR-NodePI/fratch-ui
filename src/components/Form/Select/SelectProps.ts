export type SelectOption<T> = {
  label: string;
  labelElement?: React.ReactNode;
  value: T;
  visible?: boolean;
};

export interface SelectProps<T> {
  className?: string;
  cleanable?: boolean;
  disabled?: boolean;
  noResultsElement?: React.ReactNode;
  onChange?: (value?: T) => void;
  options: SelectOption<T>[];
  placeholder?: string;
  searchable?: boolean;
  value?: T;
  triggerElementRef?: React.RefObject<HTMLInputElement>;
}
