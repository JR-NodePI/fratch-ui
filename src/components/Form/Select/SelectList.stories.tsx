import Select from './Select';
import type { Meta, StoryObj } from '@storybook/react';

const SelectList = () => {
  return (
    <>
      <h1>Select component list</h1>
      <p>
        In order to test the zindex between the select and the options list, we need to have more than one
        select in the page.
      </p>
      <ul>
        {Array.from(Array(4).keys()).map(index => {
          return (
            <li key={index} style={{ width: `${25 * (index + 1)}%` }}>
              <Select
                placeholder={`Instance ${index + 1}...`}
                options={Array.from(Array(25).keys()).map(i => ({
                  label: `Option ${index} ${i + 1} label`,
                  value: { opt: `opt-${index}-${i + 1}`, prop: `prop-${index}-${i + 1}` },
                }))}
              />
            </li>
          );
        })}
      </ul>
    </>
  );
};

const meta = {
  title: 'Example/Form/SelectList',
  component: SelectList,
  tags: [],
  argTypes: { onChange: { action: 'onChange' } },
} satisfies Meta<typeof SelectList>;

export default meta;
type Story = StoryObj<typeof SelectList>;

export const Default: Story = {
  args: {},
};
