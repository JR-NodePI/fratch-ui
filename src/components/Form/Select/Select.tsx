import { useEffect, useRef, useState } from 'react';
import { createPortal } from 'react-dom';

import { isEqual } from 'lodash';

import { c } from '../../../helpers/classNameHelpers';
import { isAscendantEvenTargetByID } from '../../../helpers/htmlSelectorsHelpers';
import { IconArrowDown } from '../../Icons/Icons';
import InputText from '../InputText/InputText';
import SelectOptionsList from './SelectOptionsList';
import { type SelectOption, type SelectProps } from './SelectProps';

import styles from './Select.module.css';

const selectInstances = new Map();

function filterOptionByText<T>(text: string) {
  return (option: SelectOption<T>): SelectOption<T> => {
    const nLabel = option.label.toLowerCase();
    const nText = text.toLowerCase();
    const visible = nLabel.includes(nText);

    let labelElement;
    if (visible) {
      const indexStart = nLabel.indexOf(nText);
      const indexEnd = indexStart + nText.length;

      const textStart = option.label.slice(0, indexStart);
      const text = option.label.slice(indexStart, indexEnd);
      const textEnd = option.label.slice(indexEnd);

      labelElement = (
        <>
          <span>{textStart}</span>
          <b>{text}</b>
          <span>{textEnd}</span>
        </>
      );
    }

    return Object.assign({}, option, { visible, labelElement });
  };
}

const useHideOnClickedOutside = ({
  uid,
  setVisible,
}: {
  uid: string;
  setVisible: React.Dispatch<React.SetStateAction<boolean>>;
}): void => {
  // hide options list when clicked outside
  useEffect(() => {
    const handleDocumentClick = (event: MouseEvent): void => {
      if (!isAscendantEvenTargetByID(event, uid)) {
        setVisible(false);
      }
    };
    window.document.body.addEventListener('click', handleDocumentClick);
    return () => {
      window.document.body.removeEventListener('click', handleDocumentClick);
    };
  }, [uid, setVisible]);
};

const useHideOnTriggerDOMRectChange = ({
  setVisible,
  visible,
  triggerDOMRect,
  triggerRef,
}: {
  setVisible: React.Dispatch<React.SetStateAction<boolean>>;
  triggerDOMRect?: DOMRect;
  triggerRef: React.RefObject<HTMLInputElement>;
  visible: boolean;
}): void => {
  useEffect(() => {
    if (visible && triggerDOMRect && triggerRef.current) {
      const intervalId = setInterval(() => {
        const newTriggerDOMRect =
          triggerRef.current && triggerRef.current.getBoundingClientRect();

        if (
          visible &&
          triggerDOMRect &&
          newTriggerDOMRect &&
          (newTriggerDOMRect.top !== triggerDOMRect.top ||
            newTriggerDOMRect.left !== triggerDOMRect.left ||
            newTriggerDOMRect.width !== triggerDOMRect.width ||
            newTriggerDOMRect.height !== triggerDOMRect.height ||
            newTriggerDOMRect.bottom !== triggerDOMRect.bottom)
        ) {
          setVisible(false);
        }
      }, 10);

      return () => {
        clearInterval(intervalId);
      };
    }
  }, [setVisible, triggerDOMRect, triggerRef, visible]);
};

function useKeyboardNavigation<T>({
  filteredOptions,
  handleOnChange,
  handleOnInputClick,
  selectedIndex,
  setVisible,
  visible,
}: {
  visible: boolean;
  selectedIndex?: number;
  filteredOptions: SelectOption<T>[];
  handleOnChange: (index: number) => void;
  handleOnInputClick: (event: React.MouseEvent<HTMLInputElement>) => void;
  setVisible: React.Dispatch<React.SetStateAction<boolean>>;
}) {
  const [focusedItemIndex, setFocusedItemIndex] = useState<number>();

  useEffect(() => {
    if (!visible) {
      setFocusedItemIndex(undefined);
    }
  }, [visible]);

  useEffect(() => {
    setFocusedItemIndex(undefined);
  }, [filteredOptions]);

  const handleOnInputKeyDownCapture = (
    event: React.KeyboardEvent<HTMLInputElement>
  ): void => {
    if (
      event.code === 'ArrowDown' ||
      event.code === 'ArrowUp' ||
      event.code === 'Enter' ||
      event.code === 'Escape'
    ) {
      event.preventDefault();
    }

    if (visible) {
      const focusedIndex = focusedItemIndex ?? selectedIndex ?? -1;

      const nextIndex = filteredOptions.findIndex(
        ({ visible }, index) => visible !== false && index > focusedIndex
      );
      if (event.code === 'ArrowDown' && nextIndex >= 0) {
        setFocusedItemIndex(nextIndex);
      }

      const prevIndex = filteredOptions.reduce(
        (lastIndex, { visible }, index) => {
          if (visible !== false && index < focusedIndex && lastIndex < index) {
            return index;
          }
          return lastIndex;
        },
        -1
      );
      if (event.code === 'ArrowUp' && prevIndex >= 0) {
        setFocusedItemIndex(prevIndex);
      }

      if (event.code === 'Enter' && focusedItemIndex != null) {
        handleOnChange(focusedItemIndex);
      }

      if (event.code === 'Escape') {
        setVisible(false);
      }
    } else {
      handleOnInputClick(event as any);
    }
  };

  return { handleOnInputKeyDownCapture, focusedItemIndex, setFocusedItemIndex };
}

function Select<T>({
  className = '',
  cleanable = false,
  disabled = false,
  id,
  noResultsElement,
  onChange,
  onClean,
  options,
  placeholder = '',
  searchable = false,
  triggerElementRef,
  value,
}: SelectProps<T>): JSX.Element {
  const [uid] = useState<string>(crypto.randomUUID());
  const newTriggerRef = useRef<HTMLInputElement>(null);
  const triggerRef = triggerElementRef ?? newTriggerRef;
  const [selectedLabel, setSelectedLabel] = useState<string>('');
  const [selectedIndex, setSelectedIndex] = useState<number>();
  const [visible, setVisible] = useState<boolean>(false);
  const [triggerDOMRect, setTriggerDOMRect] = useState<DOMRect>();
  const [filteredOptions, setFilteredOptions] =
    useState<SelectOption<T>[]>(options);

  useEffect(() => {
    selectInstances.set(uid, {
      setVisible,
    });
    return () => {
      selectInstances.delete(uid);
    };
  }, [uid]);

  // set filtered options when options changes
  useEffect(() => {
    setFilteredOptions(options);
  }, [options]);

  // set selected label and index when value changes
  useEffect(() => {
    const index = options.findIndex(option => isEqual(option.value, value));
    if (index >= 0) {
      const { label } = options[index];
      setSelectedLabel(label);
      setSelectedIndex(index);
    }
  }, [value, options]);

  const handleOnChange = (index: number): void => {
    const { value, label } = options[index];
    setSelectedLabel(label);
    setSelectedIndex(index);
    setVisible(false);
    onChange && onChange(value);

    setTimeout(() => {
      setFilteredOptions(options);
    }, 1000);
  };

  const handleOnClean = (): void => {
    onClean?.();
    if (selectedLabel) {
      setSelectedLabel('');
      setSelectedIndex(undefined);
      onChange && onChange(undefined);
      setFilteredOptions(options);
    }
  };

  const handleOnInputChange = (
    event?: React.ChangeEvent<HTMLInputElement>
  ): void => {
    const text = event?.target?.value ?? '';
    setSelectedLabel(text);
    setSelectedIndex(undefined);

    if (text.trim() === '') {
      setFilteredOptions(options);
    } else {
      const data = options.map(filterOptionByText<T>(text));
      setFilteredOptions(data);
      setVisible(true);
    }
  };

  const handleOnInputBlur = (): void => {
    if (selectedIndex != null && selectedIndex >= 0) {
      const { label } = options[selectedIndex];
      setSelectedLabel(label);
    }
  };

  const handleOnInputClick = (): void => {
    if (!disabled && triggerRef.current) {
      // hide other select instances
      selectInstances.forEach((instance, key) => {
        if (key !== uid) {
          instance.setVisible(false);
        }
      });

      setVisible(true);
      setTriggerDOMRect(triggerRef.current.getBoundingClientRect());
    }
  };

  const { focusedItemIndex, setFocusedItemIndex, handleOnInputKeyDownCapture } =
    useKeyboardNavigation({
      filteredOptions,
      handleOnChange,
      handleOnInputClick,
      selectedIndex,
      setVisible,
      visible,
    });

  // hide options list when disabled
  useEffect(() => {
    if (disabled) {
      setVisible(false);
    }
  }, [disabled]);

  useHideOnClickedOutside({ uid, setVisible });

  useHideOnTriggerDOMRectChange({
    setVisible,
    triggerDOMRect,
    triggerRef,
    visible,
  });

  return (
    <div
      id={uid}
      className={c(className, styles.select, disabled ? styles.disabled : '')}
    >
      <div className={c(styles.controls)}>
        <div>
          <InputText
            id={id}
            title={selectedLabel}
            ref={triggerRef}
            className={c(styles.trigger)}
            cleanerClassName={c(styles.cleaner)}
            disabled={disabled}
            onChange={handleOnInputChange}
            onClean={handleOnClean}
            onBlur={handleOnInputBlur}
            onKeyDownCapture={handleOnInputKeyDownCapture}
            onClick={handleOnInputClick}
            placeholder={placeholder}
            readOnly={!searchable}
            value={selectedLabel}
            cleanable={cleanable}
          />
          <IconArrowDown className={c(styles.trigger_icon)} />
        </div>
      </div>

      {createPortal(
        <SelectOptionsList
          noResultsElement={noResultsElement}
          onChange={handleOnChange}
          options={filteredOptions}
          selectedIndex={selectedIndex}
          focusedIndex={focusedItemIndex}
          setFocusedItemIndex={setFocusedItemIndex}
          triggerDOMRect={triggerDOMRect}
          visible={visible}
        />,
        document.body,
        uid
      )}
    </div>
  );
}

export default Select;
