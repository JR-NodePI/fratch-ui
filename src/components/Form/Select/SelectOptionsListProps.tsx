import { SelectOption } from './SelectProps';

export type SelectOptionsListProps<T> = {
  noResultsElement?: React.ReactNode;
  onChange?: (value: number) => void;
  options: SelectOption<T>[];
  selectedIndex?: number;
  focusedIndex?: number;
  setFocusedItemIndex: (index: number) => void;
  triggerDOMRect?: DOMRect;
  visible?: boolean;
};
