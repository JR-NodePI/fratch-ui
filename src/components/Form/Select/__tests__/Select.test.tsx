import React from 'react';

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';

import Select from '../Select';
import type { SelectOption, SelectProps } from '../SelectProps';

describe('Select', () => {
  type TestOption = {
    id: string;
    label: string;
  };

  const options: SelectOption<TestOption>[] = [
    {
      label: 'Option 1',
      value: { id: '1', label: 'Option 1' },
    },
    {
      label: 'Option 2',
      value: { id: '2', label: 'Option 2' },
      visible: false,
    },
    {
      label: 'Option 3',
      value: { id: '3', label: 'Option 3' },
    },
    {
      label: 'Option 4',
      value: { id: '4', label: 'Option 4' },
    },
  ];

  const initialProps: SelectProps<TestOption> = {
    className: 'test-class',
    cleanable: true,
    searchable: true,
    id: 'mock-test-id',
    placeholder: 'Select an option',
    options,
    value: options[0].value,
  };

  const setup = (props = initialProps) => render(<Select {...props} />);

  beforeEach(() => {
    vi.spyOn(crypto, 'randomUUID').mockReturnValue(
      'crypto-mock-random-UUID-1234'
    );
  });

  afterEach(() => {
    vi.resetAllMocks();
  });

  it('should render properly', () => {
    const { baseElement } = setup();
    expect(baseElement).toMatchSnapshot();
  });

  it('should render as visible on trigger click', async () => {
    const { baseElement } = setup();
    await userEvent.click(screen.getByRole('textbox', { name: /Option 1/i }));
    expect(baseElement).toMatchSnapshot();
  });

  it('should render as disabled', () => {
    const { container } = setup({
      ...initialProps,
      disabled: true,
    });
    expect(container).toMatchSnapshot();
  });

  it('should render the options list displayer by ref', async () => {
    const triggerElementRef = React.createRef<HTMLInputElement>();
    setup({
      ...initialProps,
      triggerElementRef,
    });

    expect(triggerElementRef).toMatchSnapshot();
  });

  it('should perform onChange', async () => {
    const onChange = vi.fn();
    setup({
      ...initialProps,
      onChange,
    });

    await userEvent.click(screen.getByRole('link', { name: /Option 3/i }));

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith({ id: '3', label: 'Option 3' });
  });

  it('should render a filtered list of options', async () => {
    const { baseElement } = setup({
      ...initialProps,
      value: undefined,
    });
    const triggerElement = screen.getByRole('textbox');
    await userEvent.type(triggerElement, 'Option 3');
    expect(baseElement).toMatchSnapshot();
  });

  it('should perform onClean and clean filtered options', async () => {
    const onClean = vi.fn();
    const onChange = vi.fn();
    const { baseElement } = setup({
      ...initialProps,
      onClean,
      onChange,
    });

    await userEvent.click(screen.getByRole('button', { name: /clean/i }));

    expect(onClean).toHaveBeenCalledTimes(1);
    expect(onClean).toHaveBeenCalledWith();

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith(undefined);

    expect(baseElement).toMatchSnapshot();
  });
});
