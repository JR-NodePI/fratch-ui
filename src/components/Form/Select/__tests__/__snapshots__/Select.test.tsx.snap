// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Select > should perform onClean and clean filtered options 1`] = `
<body>
  <div>
    <div
      class="test-class _select_bae547"
      id="crypto-mock-random-UUID-1234"
    >
      <div
        class="_controls_bae547"
      >
        <div>
          <div
            class="_input_text_300ab4 _trigger_bae547"
          >
            <input
              class="_input_300ab4"
              id="mock-test-id"
              name="mock-test-id"
              placeholder="Select an option"
              title=""
              type="text"
            />
            <button
              class="_cleaner_300ab4 _cleaner_bae547"
            >
              <svg
                aria-hidden="true"
                class="_icon_2be8fc _primary_2be8fc _cleaner_icon_300ab4"
                fill="none"
                viewBox="-1 -1 16 16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m1 1 6 6m0 0 6 6M7 7l6-6M7 7l-6 6"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
              <i>
                Clean
              </i>
            </button>
          </div>
          <svg
            aria-hidden="true"
            class="_icon_2be8fc _primary_2be8fc _trigger_icon_bae547"
            fill="none"
            viewBox="-1 -4 16 16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m1 1 5.326 5.7a.909.909 0 0 0 1.348 0L13 1"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  <div
    class="_select_list_703b77 _hidden_703b77"
    style="left: 0px; width: 0px; top: 0px; max-height: 0;"
  >
    <ul>
      <li
        class=""
        data-index="0"
      >
        <a
          data-index="0"
          href="#"
          title="Option 1"
        >
          Option 1
        </a>
      </li>
      <li
        class=""
        data-index="2"
      >
        <a
          data-index="2"
          href="#"
          title="Option 3"
        >
          Option 3
        </a>
      </li>
      <li
        class=""
        data-index="3"
      >
        <a
          data-index="3"
          href="#"
          title="Option 4"
        >
          Option 4
        </a>
      </li>
    </ul>
  </div>
</body>
`;

exports[`Select > should render a filtered list of options 1`] = `
<body>
  <div>
    <div
      class="test-class _select_bae547"
      id="crypto-mock-random-UUID-1234"
    >
      <div
        class="_controls_bae547"
      >
        <div>
          <div
            class="_input_text_300ab4 _trigger_bae547"
          >
            <input
              class="_input_300ab4"
              id="mock-test-id"
              name="mock-test-id"
              placeholder="Select an option"
              title="Option 3"
              type="text"
            />
            <button
              class="_cleaner_300ab4 _cleaner_bae547"
            >
              <svg
                aria-hidden="true"
                class="_icon_2be8fc _primary_2be8fc _cleaner_icon_300ab4"
                fill="none"
                viewBox="-1 -1 16 16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m1 1 6 6m0 0 6 6M7 7l6-6M7 7l-6 6"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
              <i>
                Clean
              </i>
            </button>
          </div>
          <svg
            aria-hidden="true"
            class="_icon_2be8fc _primary_2be8fc _trigger_icon_bae547"
            fill="none"
            viewBox="-1 -4 16 16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m1 1 5.326 5.7a.909.909 0 0 0 1.348 0L13 1"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  <div
    class="_select_list_703b77 _visible_703b77"
    style="left: 0px; width: 0px; top: 0px; max-height: 0;"
  >
    <ul>
      <li
        class=""
        data-index="2"
      >
        <a
          data-index="2"
          href="#"
          title="Option 3"
        >
          <span />
          <b>
            Option 3
          </b>
          <span />
        </a>
      </li>
    </ul>
  </div>
</body>
`;

exports[`Select > should render as disabled 1`] = `
<div>
  <div
    class="test-class _select_bae547 _disabled_bae547"
    id="crypto-mock-random-UUID-1234"
  >
    <div
      class="_controls_bae547"
    >
      <div>
        <div
          class="_input_text_300ab4 _trigger_bae547"
        >
          <input
            class="_input_300ab4"
            disabled=""
            id="mock-test-id"
            name="mock-test-id"
            placeholder="Select an option"
            title="Option 1"
            type="text"
          />
          <button
            class="_cleaner_300ab4 _cleaner_bae547"
          >
            <svg
              aria-hidden="true"
              class="_icon_2be8fc _primary_2be8fc _cleaner_icon_300ab4"
              fill="none"
              viewBox="-1 -1 16 16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m1 1 6 6m0 0 6 6M7 7l6-6M7 7l-6 6"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
            <i>
              Clean
            </i>
          </button>
        </div>
        <svg
          aria-hidden="true"
          class="_icon_2be8fc _primary_2be8fc _trigger_icon_bae547"
          fill="none"
          viewBox="-1 -4 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m1 1 5.326 5.7a.909.909 0 0 0 1.348 0L13 1"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`Select > should render as visible on trigger click 1`] = `
<body>
  <div>
    <div
      class="test-class _select_bae547"
      id="crypto-mock-random-UUID-1234"
    >
      <div
        class="_controls_bae547"
      >
        <div>
          <div
            class="_input_text_300ab4 _trigger_bae547"
          >
            <input
              class="_input_300ab4"
              id="mock-test-id"
              name="mock-test-id"
              placeholder="Select an option"
              title="Option 1"
              type="text"
            />
            <button
              class="_cleaner_300ab4 _cleaner_bae547"
            >
              <svg
                aria-hidden="true"
                class="_icon_2be8fc _primary_2be8fc _cleaner_icon_300ab4"
                fill="none"
                viewBox="-1 -1 16 16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m1 1 6 6m0 0 6 6M7 7l6-6M7 7l-6 6"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
              <i>
                Clean
              </i>
            </button>
          </div>
          <svg
            aria-hidden="true"
            class="_icon_2be8fc _primary_2be8fc _trigger_icon_bae547"
            fill="none"
            viewBox="-1 -4 16 16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m1 1 5.326 5.7a.909.909 0 0 0 1.348 0L13 1"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  <div
    class="_select_list_703b77 _visible_703b77"
    style="left: 0px; width: 0px; top: 0px; max-height: 0;"
  >
    <ul>
      <li
        class="_selected_703b77"
        data-index="0"
      >
        <a
          data-index="0"
          href="#"
          title="Option 1"
        >
          Option 1
        </a>
      </li>
      <li
        class=""
        data-index="2"
      >
        <a
          data-index="2"
          href="#"
          title="Option 3"
        >
          Option 3
        </a>
      </li>
      <li
        class=""
        data-index="3"
      >
        <a
          data-index="3"
          href="#"
          title="Option 4"
        >
          Option 4
        </a>
      </li>
    </ul>
  </div>
</body>
`;

exports[`Select > should render properly 1`] = `
<body>
  <div>
    <div
      class="test-class _select_bae547"
      id="crypto-mock-random-UUID-1234"
    >
      <div
        class="_controls_bae547"
      >
        <div>
          <div
            class="_input_text_300ab4 _trigger_bae547"
          >
            <input
              class="_input_300ab4"
              id="mock-test-id"
              name="mock-test-id"
              placeholder="Select an option"
              title="Option 1"
              type="text"
            />
            <button
              class="_cleaner_300ab4 _cleaner_bae547"
            >
              <svg
                aria-hidden="true"
                class="_icon_2be8fc _primary_2be8fc _cleaner_icon_300ab4"
                fill="none"
                viewBox="-1 -1 16 16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m1 1 6 6m0 0 6 6M7 7l6-6M7 7l-6 6"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
              <i>
                Clean
              </i>
            </button>
          </div>
          <svg
            aria-hidden="true"
            class="_icon_2be8fc _primary_2be8fc _trigger_icon_bae547"
            fill="none"
            viewBox="-1 -4 16 16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m1 1 5.326 5.7a.909.909 0 0 0 1.348 0L13 1"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  <div
    class="_select_list_703b77 _hidden_703b77"
    style="left: 0px; width: 0px; top: 0px; max-height: 0;"
  >
    <ul>
      <li
        class="_selected_703b77"
        data-index="0"
      >
        <a
          data-index="0"
          href="#"
          title="Option 1"
        >
          Option 1
        </a>
      </li>
      <li
        class=""
        data-index="2"
      >
        <a
          data-index="2"
          href="#"
          title="Option 3"
        >
          Option 3
        </a>
      </li>
      <li
        class=""
        data-index="3"
      >
        <a
          data-index="3"
          href="#"
          title="Option 4"
        >
          Option 4
        </a>
      </li>
    </ul>
  </div>
</body>
`;

exports[`Select > should render the options list displayer by ref 1`] = `
{
  "current": <input
    class="_input_300ab4"
    id="mock-test-id"
    name="mock-test-id"
    placeholder="Select an option"
    title="Option 1"
    type="text"
  />,
}
`;
